tools/state-viewer/src/cli.rs

```diff
@@ -606,3 +620,214 @@ impl ViewTrieCmd {
+    #[clap(long, default_value = "full")]
+    iteration_type: TrieIterationType,
+
+    /// Limit the number of trie nodes to be iterated.
+    #[clap(long)]
+    limit: Option<u32>,
+
+    /// Print the trie nodes to stdout.
+    #[clap(long, default_value = "false")]
+    print: bool,
+}
+
+impl TrieIterationBenchmarkCmd {
+    pub fn run(self, _home_dir: &Path, near_config: NearConfig, store: Store) {
+        let genesis_config = &near_config.genesis.config;
+        let chain_store = ChainStore::new(
+            store.clone(),
+            genesis_config.genesis_height,
+            near_config.client_config.save_trie_changes,
+        );
+        let head = chain_store.head().unwrap();
+        let block = chain_store.get_block(&head.last_block_hash).unwrap();
+        let epoch_manager =
+            EpochManager::new_from_genesis_config(store.clone(), &genesis_config).unwrap();
+        let shard_layout = epoch_manager.get_shard_layout(block.header().epoch_id()).unwrap();
+
+        for (shard_id, chunk_header) in block.chunks().iter().enumerate() {
+            let shard_id = shard_id as ShardId;
+            if chunk_header.height_included() != block.header().height() {
+                println!("chunk for shard {shard_id} is missing");
+                return;
+            }
+            let trie = self.get_trie(shard_id, &shard_layout, &chunk_header, &store);
+
+            println!("shard id    {shard_id:#?} benchmark starting");
+            let prune_condition: Option<Box<dyn Fn(&Vec<u8>) -> bool>> =
+                Some(Box::new(Self::shallow_iter_prune_condition));
+            match &self.iteration_type {
+                TrieIterationType::Full => self.iter_trie(&trie, None),
+                TrieIterationType::Shallow => self.iter_trie(&trie, prune_condition),
+            };
+            println!("shard id    {shard_id:#?} benchmark finished");
+        }
+    }
+
+    fn get_trie(
+        &self,
+        shard_id: ShardId,
+        shard_layout: &near_primitives::shard_layout::ShardLayout,
+        chunk_header: &near_primitives::sharding::ShardChunkHeader,
+        store: &Store,
+    ) -> Trie {
+        let shard_uid = ShardUId::from_shard_id_and_layout(shard_id, shard_layout);
+        // Note: here we get the previous state root but the shard layout
+        // corresponds to the current epoch id. In practice shouldn't
+        // matter as the shard layout doesn't change.
+        let state_root = chunk_header.prev_state_root();
+        // Question for reviewer - there are different TrieStorage
+        // implementation, which would be best suited here? I picked this
+        // one since it looked simplest and I don't think there is much
+        // point in using caches.
+        let storage = TrieDBStorage::new(store.clone(), shard_uid);
+        // Question for reviewer - my assumption is that flat storgage
+        // should be disabled or is irrelevant for this task, can you
+        // confirm please?
```

Comment:
Yes, we don't need flat storage here - the trie iteration is going to skip flat storage anyway. It's only relevant if you're looking up point-wise. (At least I don't see how it would work at all otherwise).